src/dex/aave-v3/utils.ts

```diff
@@ -0,0 +1,156 @@
+import {
+  extractSuccessAndValue,
+  stringDecode,
+  uint8ToNumber,
+} from '../../lib/decoders';
+import {
+  MultiCallParams,
+  MultiResult,
+  MultiWrapper,
+} from '../../lib/multi-wrapper';
+import Web3 from 'web3';
+import POOL_ABI from '../../abi/AaveV3_lending_pool.json';
+import { AaveToken } from './types';
+
+export const decodeATokenFromReserveData = (
+  result: MultiResult<BytesLike> | BytesLike,
+): string => {
+  const [isSuccess, toDecode] = extractSuccessAndValue(result);
+
+  if (!isSuccess || toDecode === '0x') {
+    return 'error';
+  }
+
+  const decoded = defaultAbiCoder.decode(
+    [
+      'tuple(uint256) configuration',
+      'uint128 liquidityIndex',
+      'uint128 currentLiquidityRate',
+      'uint128 variableBorrowIndex',
+      'uint128 currentVariableBorrowRate',
+      'uint128 currentStableBorrowRate',
+      'uint40 lastUpdateTimestamp',
+      'uint16 id',
+      'address aTokenAddress',
+      'address stableDebtTokenAddress',
+      'address variableDebtTokenAddress',
+      'address interestRateStrategyAddress',
+      'uint128 accruedToTreasury',
+      'uint128 unbacked',
+      'uint128 isolationModeTotalDebt',
+    ],
+    toDecode,
+  );
+
+  return decoded.aTokenAddress;
+};
+
+async function _getAllTokenMetadata(
+  blockNumber: number,
+  reservesList: string[],
+  poolAddress: string,
+  poolInterface: Interface,
+  erc20Interface: Interface,
+  multiWrapper: MultiWrapper,
+): Promise<
+  {
+    address: string;
+    aAddress: string;
+    decimals: number;
+  }[]
+> {
+  let calls: MultiCallParams<any>[] = [];
+
+  for (const tokenAddress of reservesList) {
+    calls.push({
+      target: poolAddress,
+      callData: poolInterface.encodeFunctionData('getReserveData', [
+        tokenAddress,
+      ]),
+      decodeFunction: decodeATokenFromReserveData,
+    });
+  }
+  for (const proxyAddress of reservesList) {
+    calls.push({
+      target: proxyAddress,
+      callData: erc20Interface.encodeFunctionData('decimals'),
+      decodeFunction: uint8ToNumber,
+    });
+  }
+
+  const results = await multiWrapper.tryAggregate<string | number>(
+    false,
+    calls,
+    blockNumber,
+  );
+
+  let tokenList = reservesList.map((tokenAddress: string, i: number) => ({
+    address: tokenAddress as string,
+    aAddress: results[i].returnData as string,
+    decimals: results[reservesList.length + i].returnData as number,
+  }));
+
+  return tokenList;
+}
+
+async function _getATokenSymbols(
+  blockNumber: number,
+  aTokens: string[],
+  erc20Interface: Interface,
+  multiWrapper: MultiWrapper,
+): Promise<{ aSymbol: string }[]> {
+  let calls: MultiCallParams<any>[] = [];
+
+  for (const proxyAddress of aTokens) {
+    calls.push({
+      target: proxyAddress,
+      callData: erc20Interface.encodeFunctionData('symbol', []),
+      decodeFunction: stringDecode,
+    });
+  }
+  const results = await multiWrapper.tryAggregate<string>(
```

Comment:
If we don't expect any call to fail, better to use `aggregate`. It is more efficient and will fail explicitly. But only if all calls must pass

Reply:
There is no reason any call should fail actually.

Done.